tests/core/test_reader.py

```diff
@@ -64,20 +82,36 @@ def mock_sensor(mock_serial):
     return mock_serial
 
 
-def test_sensor_reader(mock_sensor, monkeypatch):
-    sensor_reader = reader.SensorReader(
-        port=mock_sensor.port,
+@pytest.fixture
+def sensor_reader_factory(monkeypatch, mock_sensor):
+    def factory(
+        *,
```

Comment:
maybe instead of implementing this fixture as a closure/factory, it would be cleaner to use [indirect parametrization]. 

On its simplest form, the fixture would look like:
``` python
@pytest.fixture
def sensor_reader(monkeypatch, mock_sensor, samples: int, interval: int | None, sensor: str):
    sensor_reader = reader.SensorReader(
        port=mock_sensor.port,
        samples=samples,
        interval=interval,
        sensor=sensor,
        timeout=0.01,  # low to avoid hanging on failure
    )

    # https://github.com/pyserial/pyserial/issues/625
    monkeypatch.setattr(sensor_reader.serial, "flush", lambda: None)

    return sensor_reader
```

and the test function would look like
``` python
@pytest.mark.parameterize("samples,interval,sensor",
    [
        (0, None, "PMSx003")  # exit immediately, match with stubs
    ]
) 
def test_sensor_reader(mock_sensor, sensor_reader: reader.Reader):
    [...]
```

As far as I know, it is not possible to define optional arguments when using indirect parameterization, so every test needs to define all its inputs.

Something I particularly like of this method is that allows to keep all the 
test setup on the fixture.

[indirect parametrization]: https://docs.pytest.org/en/latest/example/parametrize.html#apply-indirect-on-particular-arguments

Reply:
This is interesting: I've not heard of indirect parametrisation before. It was annoying when I first introduced the fixture and then had to convert it to a factory to make it configurable just for some of the tests.

I'm wondering if using this indirect parameterisation would cause more problems than it solves though:

- We lose some of the DRY effect of the last commit - by having to repeat all the optional parameters. This means there's actually more setup in most of the tests than with the factory approach.

- Let's say I want to add a future test that needs to override `timeout`. Then I have to go and edit every other test to define the `timeout` parameter, as opposed to with the current factory approach.

- Indirect parameterisation could be confusing to someone who doesn't know about it - it's very...indirect 😉. Maybe I'm in a minority with not knowing it, but I think it's still a valid concern.

So I'm not convinced this would be an improvement - sometimes I find a bit of setup is just clearer than a lot of abstraction. Your call though - I'll give it a go if it's the approach you want to take here.

Reply:
The test work and are readable, so lets keep the factory. I might refactor it at a later stage, but that is future...